### PR TITLE
Add the retry logic to the request to get metadata in MetadataFetcher.

### PR DESCRIPTION
**Pull request by siqidong**

### CODE REVIEW

**Comment by siqidong** [comment]
Hi @TAOXUY @shuoyang2016 Can you help to review this PR? Thank you!

**Comment by TAOXUY** [comment]
Test the retry behavior in the `metadata_fetcher_test.go`?

File: src/go/metadata/metadata_fetcher_test.go [hunk 1]
```diff
@@ -310,6 +312,12 @@ func TestFetchGCPAttributes(t *testing.T) {
 		mf := NewMockMetadataFetcher(mockBaseUrl, time.Now())
 
 		attrs, err := mf.FetchGCPAttributes()
+		// if error is thrown, actual error should be equal to the expectedError
+		if err != nil && err.Error() != tc.expectedError {
+			t.Errorf(errorTmpl, tc.desc, tc.expectedError, err)
```
Comment (TAOXUY) [inline src/go/metadata/metadata_fetcher_test.go:317]: 1. the template should be `expected error` and `actual error`
2. it miss the case of err == nil && tc.expectedError != ""

Comment (siqidong) [reply src/go/metadata/metadata_fetcher_test.go]: Thank you Jason!
1. Renamed the error naming to be `expectedError` and `actualError`.
2. Modified the test case to include the situation when actual error is nil but expected error is not nil.

File: src/go/metadata/metadata_fetcher_test.go [hunk 2]
```diff
@@ -309,9 +311,15 @@ func TestFetchGCPAttributes(t *testing.T) {
 
 		mf := NewMockMetadataFetcher(mockBaseUrl, time.Now())
 
-		attrs, err := mf.FetchGCPAttributes()
-		if err != nil && tc.expectedGCPAttributes != nil {
-			t.Errorf(errorTmpl, tc.desc, nil, err)
+		attrs, actualError := mf.FetchGCPAttributes()
+		// if expected error is set, actual error should be equal to the expected error
+		if tc.expectedError != "" && actualError.Error() != tc.expectedError {
+			t.Errorf(errorTmpl, tc.desc, tc.expectedError, actualError)
```
Comment (TAOXUY) [inline src/go/metadata/metadata_fetcher_test.go:317]: The error should be `Test: %s\n ExpectedError: %v\n ActualError: %v"`

Comment (siqidong) [reply src/go/metadata/metadata_fetcher_test.go:317]: Line 302 defines `errorTmpl := "Test: %s\n  Expected: %v\n  Actual: %v"`, which is a general statement that already covers the error case.

Comment (TAOXUY) [reply src/go/metadata/metadata_fetcher_test.go:317]: It isn't cleared about the mismatch is for the expected resp or the expected error

Comment (siqidong) [reply src/go/metadata/metadata_fetcher_test.go:318]: Sure, added `attrsTmpl` to distinguish the attrs and error templates. Thanks.

File: src/go/metadata/metadata_fetcher_test.go [hunk 3]
```diff
@@ -309,22 +312,28 @@ func TestFetchGCPAttributes(t *testing.T) {
 
 		mf := NewMockMetadataFetcher(mockBaseUrl, time.Now())
 
-		attrs, err := mf.FetchGCPAttributes()
-		if err != nil && tc.expectedGCPAttributes != nil {
-			t.Errorf(errorTmpl, tc.desc, nil, err)
+		attrs, actualError := mf.FetchGCPAttributes()
+		// if expected error is set, actual error should be equal to the expected error
```
Comment (TAOXUY) [inline src/go/metadata/metadata_fetcher_test.go:316]: * after second thought, please move this into a separate test like `TestFetchGCPAttributes_Errors` go/gotip/episodes/50
* make the retry setting part of testCase for better readability

Comment (siqidong) [reply src/go/metadata/metadata_fetcher_test.go]: > * after second thought, please move this into a separate test like `TestFetchGCPAttributes_Errors` go/gotip/episodes/50

Added individual test case `TestFetchGCPAttributes_NoMetadataServer_FailsAllRetries`.
> * make the retry setting part of testCase for better readability

Retry setting is added in `metadata_fetcher_test_util.go`, which will be shared by all test cases.

Comment (TAOXUY) [reply src/go/metadata/metadata_fetcher_test.go]: Prefer to explicitly set it in test setup so that we know what we are testing against

**Review by TAOXUY — APPROVED** [review]

**Comment by TAOXUY** [comment]
@siqidong , looks like you haven't updated its caller.

https://screenshot.googleplex.com/6DaZzePRYbx6dtW

You can run `make test` to validate the unit tests passed locally.

**Review by TAOXUY — APPROVED** [review]

**Comment by TAOXUY** [comment]
Hi April, you can send multiple commits in one PR and it allows reviewers to better review what is changed in individual commit.

**Comment by siqidong** [comment]
> Hi April, you can send multiple commits in one PR and it allows reviewers to better review what is changed in individual commit.

Thanks Jason! Will do it next time. I have fixed the integ test in my previous commit by changing the retry configs. Thanks for review!

**Comment by TAOXUY** [comment]
Great! Your PR passed the presubmit!

File: src/go/metadata/metadata_fetcher_test_util.go [hunk 4]
```diff
@@ -18,27 +18,30 @@ import (
 	"time"
 
 	"github.com/GoogleCloudPlatform/esp-v2/src/go/options"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/util"
 )
 
 // Creates a mock metadata fetcher and returns the mock instance
-func NewMockMetadataFetcher(baseUrl string, now time.Time) *MetadataFetcher {
+func NewMockMetadataFetcher(baseUrl string, now time.Time, retryConfig util.RetryConfig) *MetadataFetcher {
```
[the remark this answers is not in the document]
Comment (siqidong) [reply src/go/metadata/metadata_fetcher_test_util.go]: Updated, thanks for the suggestion!

**Pull request merged**